module/src/clients/surface_temp.c

```diff
@@ -118,8 +135,19 @@ static int ssam_temp_hwmon_read_string(struct device *dev,
 	if (attr != hwmon_temp_label)
 		return -EOPNOTSUPP;
 
-	// TODO: get label from SAM
-	*str = "<not yet implemented>";
+	// Get the label from the SSAM if we haven't retrieved it yet.
+	if (ssam_temp->names[channel][0] == 0)
+	{
+		status =  ssam_retry(__ssam_tmp_get_name, ssam_temp->sdev->ctrl,
+				     SSAM_SSH_TID_SAM, channel + 1, &name_resp);
+		if (status < 0)
+			return -EIO;
+
+		// Store the name in the internal buffer.
+		memcpy(ssam_temp->names[channel], name_resp.name,
+		       sizeof(ssam_temp->names[channel]));
```

Comment:
There should be some string copy function for that. Specifically, one that ensures it's properly null-terminated.

Reply:
The function I was trying to remember is `strscpy(dest, src, max_count)`. Also use the macro instead of the sizeof for the third parameter.

Reply:
Done, I'm not sure if `WARN_ON` is the correct thing though, the [docs](https://github.com/linux-surface/kernel/blob/8cd23b1bb3a8b7a3ef7cec2c37e7e46e6397a858/include/asm-generic/bug.h#L75-L77)

> * WARN(), WARN_ON(), WARN_ON_ONCE, and so on can be used to report
> * significant kernel issues that need prompt attention if they should ever
> * appear at runtime.

I'm not really sure if this needs prompt attention, or whether the user can even do anything about it.

Reply:
Late reply, but I mostly use `WARN_ON` to catch bugs in the code / validate hard assumptions that we have regardless of impact/severity because it produces a nice (and somewhat harder to miss) stack trace in the log. Which I hope kind of tells the end-user "hey something went wrong here, please report this" instead of something more subtle that's easier to miss (or ignore). Apart from that I think there's generally not much the end-user can do about any `WARN_ON` being triggered. If there was, it should probably be an error-code return to user-space and not `WARN_ON`.

Reply:
:+1:  that's a good rule of thumb. Thanks for sharing!